mono/metadata/boehm-gc.c

```diff
@@ -1941,12 +1949,90 @@ mono_gc_register_obj_with_weak_fields (void *obj)
 	g_error ("Weak fields not supported by boehm gc");
 }
 
+struct ephemeron_node
+{
+	ephemeron_node* next;
+	void* ephemeron_array_weak_link;
+};
+
+
+static gpointer
+ephemeron_array_add (gpointer arg)
+{
+	ephemeron_node* item = (ephemeron_node*)arg;
+	ephemeron_node* current = ephemeron_list;
+	item->next = current;
+	ephemeron_list = item;
+
+	return NULL;
+}
+
 gboolean
 mono_gc_ephemeron_array_add (MonoObject *obj)
 {
+	ephemeron_node* item = GC_MALLOC (sizeof (ephemeron_node));
+	memset (item, 0, sizeof (ephemeron_node));
+
+	mono_gc_weak_link_add (&item->ephemeron_array_weak_link, obj, FALSE);
+
+	GC_call_with_alloc_lock (ephemeron_array_add, item);
 	return TRUE;
 }
 
+typedef struct {
+	MonoObject* key;
+	MonoObject* value;
+} Ephemeron;
+
+static void
+mono_push_ephemerons (void)
+{
+	ephemeron_node* prev = NULL;
+	ephemeron_node* current = NULL;
+
+	/* iterate all registered Ephemeron[] */
+	for (current = ephemeron_list; current; current = current->next)
+	{
+		Ephemeron* cur, * array_end;
+		MonoObject* tombstone;
```

Comment:
Is there any reason not to initialize `tombstone` where it is declared here? We could move this below the following line and always initialize `tombstone` immediately.